lib/sensors/GPS.h

```diff
@@ -0,0 +1,362 @@
+const uint8_t UBX_NAV_SIG = 0x43;       //Signal Information
+const uint8_t UBX_NAV_STATUS = 0x03;    //Receiver Navigation Status
+const uint8_t UBX_NAV_SVIN = 0x3B;      //Survey-in data. Used for checking Survey In status
+const uint8_t UBX_NAV_TIMEBDS = 0x24;   //BDS Time Solution
+const uint8_t UBX_NAV_TIMEGAL = 0x25;   //Galileo Time Solution
+const uint8_t UBX_NAV_TIMEGLO = 0x23;   //GLO Time Solution
+const uint8_t UBX_NAV_TIMEGPS = 0x20;   //GPS Time Solution
+const uint8_t UBX_NAV_TIMELS = 0x26;    //Leap second event information
+const uint8_t UBX_NAV_TIMEUTC = 0x21;   //UTC Time Solution
+const uint8_t UBX_NAV_VELECEF = 0x11;   //Velocity Solution in ECEF
+const uint8_t UBX_NAV_VELNED = 0x12;    //Velocity Solution in NED
+
+//The following are used to configure the RXM UBX messages
+const uint8_t UBX_RXM_MEASX = 0x14; //Satellite Measurements for RRLP
+const uint8_t UBX_RXM_PMREQ = 0x41; //Requests a Power Management task (two differenent packet sizes)
+const uint8_t UBX_RXM_RAWX = 0x15;  //Multi-GNSS Raw Measurement Data
+const uint8_t UBX_RXM_RLM = 0x59;   //Galileo SAR Short-RLM report (two different packet sizes)
+const uint8_t UBX_RXM_RTCM = 0x32;  //RTCM input status
+const uint8_t UBX_RXM_SFRBX = 0x13; //Boradcast Navigation Data Subframe
+
+
+//ACK ID constants
+const uint8_t UBX_ACK_NACK = 0x00;
+const uint8_t UBX_ACK_ACK = 0x01;
+
+//Port types and ID's
+const uint8_t COM_PORT_I2C = 0;
+const uint8_t COM_PORT_UART1 = 1;
+const uint8_t COM_PORT_UART2 = 2;
+const uint8_t COM_PORT_USB = 3;
+const uint8_t COM_PORT_SPI = 4;
+
+const uint8_t COM_TYPE_UBX = (1 << 0);
+const uint8_t COM_TYPE_NMEA = (1 << 1);
+const uint8_t COM_TYPE_RTCM3 = (1 << 5);
+
+typedef struct {
+  uint8_t cls;
+  uint8_t id;
+  uint16_t len;           //Length of the payload. Does not include cls, id, or checksum bytes
+  uint16_t counter;       //Keeps track of number of overall bytes received. Some responses are larger than 255 bytes.
+  uint16_t startingSpot; //The counter value needed to go past before we begin recording into payload array
+  uint8_t *payload;      // array containing payload data
+  uint8_t checksumA;      //Given to us from module. Checked against the rolling calculated A/B checksums.
+  uint8_t checksumB;
+  ublox_packet_validity_e valid; //Goes from NOT_DEFINED to VALID or NOT_VALID when checksum is checked
+  // Goes from NOT_DEFINED to VALID or NOT_VALID when the Class and ID match the requestedClass and requestedID
+  ublox_packet_validity_e classAndIDmatch;
+} ubxPacket;
+
+//Depending on the sentence type the processor will load characters into different arrays
+enum SentenceTypes {
+  NONE = 0,
+  NMEA,
+  UBX,
+  RTCM
+} currentSentence = NONE;
+
+class GPS {
+ public:
+  GPS(PinName tx, PinName rx, uint32_t baud_rate);
+  void configure(); //configures the GPS (note that serial loop must be started before this is run)
+  void saveConfiguration(); //saves configuration to persistent memory
+  void process();
```

Comment:
What is this processing? when do you call it?

Reply:
It processes one byte from the buffer and will be called in the main update loop thread for the GPS, the name could be clearer

Reply:
ok pls better name and comment